src/toolchain.ml

```diff
@@ -184,6 +184,58 @@ let of_settings () : t option Promise.t =
   | Some Global -> Promise.return (Some Global)
   | Some (Custom template) -> Promise.return (Some (Custom template))
 
+let detect_esy_sandbox ~project_root esy () =
+  let open Promise.Syntax in
+  let* esy = esy in
+  let* esy_build_dir_exists =
+    Fs.exists Path.(project_root / "_esy" |> Path.to_string)
+  in
+  let+ manifest = Esy.find_manifest_in_dir project_root in
+  match (esy, esy_build_dir_exists, manifest) with
+  | Some esy, true, Some _ ->
+    (* Some projects have both use both Esy and Opam, we can't assume that the user
+       wants to use Esy just by looking for an "esy.json" file.
+
+       Instead, we assume the user wants to use esy if these three conditions are met:
+        - The esy command is in the user's system
+        - There is an _esy build directory in the project root
+        - There is an Esy manifest file in the project root
+    *)
+    Some (Esy (esy, project_root))
+  | _ -> None
+
+let detect_opam_sandbox ~project_root opam () =
+  let open Promise.Option.Syntax in
+  let* opam = opam in
+  let+ switch = Opam.switch_show ~cwd:project_root opam in
+  Opam (opam, switch)
+
+let detect () =
+  match Workspace.workspaceFolders () with
+  | [] -> Promise.return None
+  | [ workspace_folder ] -> (
+    let project_root =
+      workspace_folder |> WorkspaceFolder.uri |> Uri.path |> Path.of_string
+    in
+    let available = available_toolchains () in
+    try
+      Promise.List.find_map
+        (fun f -> f ())
+        [ detect_esy_sandbox ~project_root available.esy
+        ; detect_opam_sandbox ~project_root available.opam
+        ]
+    with Assert_failure _ -> Promise.return None )
```

Comment:
Looks like catching `Assert_failure _` is only required because of a bug in promise_jsoo. I've sent a PR to fix it here https://github.com/mnxn/promise_jsoo/pull/1

So let's wait until @mnxn fixes and sends a release.

Reply:
Next release should be available soon: https://github.com/ocaml/opam-repository/pull/17717

Reply:
We'll need a constraint to reflect this as well.